File: ast/declaration/enum.go
package declaration

import (
	"fmt"
	"strconv"

	"github.com/panda-io/micro-panda/ast/core"
	"github.com/panda-io/micro-panda/ast/expression"
	"github.com/panda-io/micro-panda/token"
)

type Enum struct {
	DeclarationBase
	Members []*Variable
	Values  []uint8
}

func (e *Enum) IsConstant() bool {
	return false
}

func (e *Enum) Kind() core.DeclarationKind {
	return core.DeclarationEnum
}

func (e *Enum) Type() core.Type {
	return nil
}

func (e *Enum) AddMember(m *Variable) error {
	if e.HasMember(m.Name.Name) {
		return fmt.Errorf("%s redeclared", m.Name.Name)
	}
	e.Members = append(e.Members, m)
	return nil
}

func (e *Enum) HasMember(name string) bool {
	for _, v := range e.Members {
		if v.Name.Name == name {
			return true
		}
	}
	return false
}

func (e *Enum) ResolveType(c core.Context) {
}

func (e *Enum) Validate(c core.Context) {
	var index int
	for _, v := range e.Members {
		if index >= 256 {
			c.Error(v.GetPosition(), "enum value shoud be less than 256")
		}
		if v.Value == nil {
			e.Values = append(e.Values, uint8(index))
			index++
		} else {
			if literal, ok := v.Value.(*expression.Literal); ok && literal.Token == token.INT {
				if i, _ := strconv.Atoi(literal.Value); i >= index {
					index = i
					e.Values = append(e.Values, uint8(index))
					index++
				} else {
					c.Error(v.GetPosition(), fmt.Sprintf("enum value here should be greater than %d.", i-1))
				}
			} else {
				c.Error(v.GetPosition(), "enum value must be const integer.")
			}
		}
	}
}

File: MicroPanda/AST/Declaration/Enum.cs
namespace MicroPanda.AST.Declaration;

using Expression;

internal class Enum : Declaration
{
    internal List<Variable> Members = [];
    internal List<byte> Values = [];

    internal bool AddMember(Variable member)
    {
        if (HasMember(member.Name!.Name!))
        {
            throw new Exception($"{member.Name!.Name!} redeclared");
        }
        Members.Add(member);
        return true;
    }

    internal bool HasMember(string name)
    {
        foreach (var member in Members)
        {
            if (member.Name!.Name! == name)
            {
                return true;
            }
        }
        return false;
    }

    internal override void ValidateType(Context context)
    {
    }

    internal override void Validate(Context context)
    {
        int index = 0;
        foreach (var member in Members)
        {
            if (index >= 256)
            {
                context.Program.Error(member.Position, "enum value should be less than 256");
            }
            if (member.Value == null)
            {
                Values.Add((byte)index);
                index++;
            }
            else
            {
                if (member.Value is Literal literal)
                {
                    var i = literal.AsInt(context);
                    if (i >= index)
                    {
                        index = i;
                        Values.Add((byte)index);
                        index++;
                    }
                    else
                    {
                        context.Program.Error(member.Position, $"enum value here should be greater than {i - 1}.");
                    }
                }
                else
                {
                    context.Program.Error(member.Position, "enum value must be const integer.");
                }
            }
        }
    }
}